File: CDRTranslucentSideBar/CDRTranslucentSideBar/CDRTranslucentSideBar.m
//
//  CDRTranslucentSideBar.m
//  CDRTranslucentSideBar
//
//  Created by UetaMasamichi on 2014/06/16.
//  Copyright (c) 2014年 nscallop. All rights reserved.
//

#import "CDRTranslucentSideBar.h"

#define SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(v) ([[[UIDevice currentDevice] systemVersion] compare:v options:NSNumericSearch] != NSOrderedAscending)

@interface CDRTranslucentSideBar ()
@property (nonatomic, strong) UIToolbar *translucentView;
@property (nonatomic, strong) UIView *contentView;
@property (nonatomic, strong) UITapGestureRecognizer *tapGestureRecognizer;
@property (nonatomic, strong) UIPanGestureRecognizer *panGestureRecognizer;
@property CGPoint panStartPoint;
@end

@implementation CDRTranslucentSideBar

- (instancetype)init {
	self = [super init];
	if (self) {
		[self initCDRTranslucentSideBar];
	}
	return self;
}

- (instancetype)initWithDirectionFromRight:(BOOL)showFromRight {
	self = [super init];
	if (self) {
		_showFromRight = showFromRight;
		[self initCDRTranslucentSideBar];
	}
	return self;
}

- (instancetype)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil {
	self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
	if (self) {
	}
	return self;
}

#pragma mark - Custom Initializer
- (void)initCDRTranslucentSideBar {
	_hasShown = NO;
	self.isCurrentPanGestureTarget = NO;

	self.sideBarWidth = 200;
	self.animationDuration = 0.25f;

	[self initTranslucentView];

	self.view.backgroundColor = [UIColor clearColor];
	self.tapGestureRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(handleTapGesture:)];
	self.tapGestureRecognizer.delegate = self;
	[self.view addGestureRecognizer:self.tapGestureRecognizer];
	self.panGestureRecognizer = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(handlePanGesture:)];
	self.panGestureRecognizer.minimumNumberOfTouches = 1;
	self.panGestureRecognizer.maximumNumberOfTouches = 1;
	[self.view addGestureRecognizer:self.panGestureRecognizer];
}

- (void)initTranslucentView {
	if (SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(@"7.0")) {
		CGRect translucentFrame =
		    CGRectMake(self.showFromRight ? self.view.bounds.size.width : -self.sideBarWidth, 0, self.sideBarWidth, self.view.bounds.size.height);
		self.translucentView = [[UIToolbar alloc] initWithFrame:translucentFrame];
		self.translucentView.frame = translucentFrame;
		self.translucentView.contentMode = _showFromRight ? UIViewContentModeTopRight : UIViewContentModeTopLeft;
		self.translucentView.clipsToBounds = YES;
		self.translucentView.barStyle = UIBarStyleDefault;
		[self.view.layer insertSublayer:self.translucentView.layer atIndex:0];
	}
}

#pragma mark - View Life Cycle
- (void)viewDidLoad {
	[super viewDidLoad];
}

- (void)didReceiveMemoryWarning {
	[super didReceiveMemoryWarning];
}

- (void)loadView {
	[super loadView];
}

#pragma mark - Layout
- (BOOL)shouldAutorotate {
	return YES;
}

- (NSUInteger)supportedInterfaceOrientations {
	return UIInterfaceOrientationMaskAll;
}

- (void)willAnimateRotationToInterfaceOrientation:(UIInterfaceOrientation)toInterfaceOrientation duration:(NSTimeInterval)duration {
	[super willAnimateRotationToInterfaceOrientation:toInterfaceOrientation duration:duration];

	if ([self isViewLoaded] && self.view.window != nil) {
		[self layoutSubviews];
	}
}

- (void)layoutSubviews {
	CGFloat x = self.showFromRight ? self.parentViewController.view.bounds.size.width - self.sideBarWidth : 0;

	if (self.contentView != nil) {
		self.contentView.frame = CGRectMake(x, 0, self.sideBarWidth, self.parentViewController.view.bounds.size.height);
	}
}

#pragma mark - Accessor
- (void)setTranslucentStyle:(UIBarStyle)translucentStyle {
	self.translucentView.barStyle = translucentStyle;
}

- (UIBarStyle)translucentStyle {
	return self.translucentView.barStyle;
}

- (void)setTranslucent:(BOOL)translucent {
	self.translucentView.translucent = translucent;
}

- (BOOL)translucent {
	return self.translucentView.translucent;
}

- (void)setTranslucentAlpha:(CGFloat)translucentAlpha {
	self.translucentView.alpha = translucentAlpha;
}

- (CGFloat)translucentAlpha {
	return self.translucentView.alpha;
}

- (void)setTranslucentTintColor:(UIColor *)translucentTintColor {
	self.translucentView.barTintColor = translucentTintColor;
}

- (UIColor *)translucentTintColor {
	return self.translucentView.barTintColor;
}

#pragma mark - Show
- (void)showInViewController:(UIViewController *)controller animated:(BOOL)animated {
	if ([self.delegate respondsToSelector:@selector(sideBar:willAppear:)]) {
		[self.delegate sideBar:self willAppear:animated];
	}

	[self addToParentViewController:controller callingAppearanceMethods:YES];
	self.view.frame = controller.view.bounds;

	CGFloat parentWidth = self.view.bounds.size.width;
	CGRect sideBarFrame = self.view.bounds;
	sideBarFrame.origin.x = self.showFromRight ? parentWidth : -self.sideBarWidth;
	sideBarFrame.size.width = self.sideBarWidth;

	if (self.contentView != nil) {
		self.contentView.frame = sideBarFrame;
	}
	sideBarFrame.origin.x = self.showFromRight ? parentWidth - self.sideBarWidth : 0;

	void (^animations)() = ^{
		if (self.contentView != nil) {
			self.contentView.frame = sideBarFrame;
		}
		self.translucentView.frame = sideBarFrame;
	};
	void (^completion)(BOOL) = ^(BOOL finished)
	{
		_hasShown = YES;
		self.isCurrentPanGestureTarget = YES;
		if (finished && [self.delegate respondsToSelector:@selector(sideBar:didAppear:)]) {
			[self.delegate sideBar:self didAppear:animated];
		}
	};

	if (animated) {
		[UIView animateWithDuration:self.animationDuration delay:0 options:kNilOptions animations:animations completion:completion];
	}
	else {
		animations();
		completion(YES);
	}
}

-(void)showInViewController:(UIViewController *)controller {
    [self showInViewController:controller animated:YES];
}

- (void)showAnimated:(BOOL)animated {
	UIViewController *controller = [UIApplication sharedApplication].keyWindow.rootViewController;
	while (controller.presentedViewController != nil) {
		controller = controller.presentedViewController;
	}
	[self showInViewController:controller animated:animated];
}

- (void)show {
	[self showAnimated:YES];
}

#pragma mark - Show by Pangesture
- (void)startShow:(CGFloat)startX {
	UIViewController *controller = [UIApplication sharedApplication].keyWindow.rootViewController;
	while (controller.presentedViewController != nil) {
		controller = controller.presentedViewController;
	}
    [self startShowInViewController:controller startX:startX];
}

-(void)startShowInViewController:(UIViewController *)controller startX:(CGFloat)startX{
    [self addToParentViewController:controller callingAppearanceMethods:YES];
    self.view.frame = controller.view.bounds;
    
    CGFloat parentWidth = self.view.bounds.size.width;
    
    CGRect sideBarFrame = self.view.bounds;
    sideBarFrame.origin.x = self.showFromRight ? parentWidth : -self.sideBarWidth;
    sideBarFrame.size.width = self.sideBarWidth;
    if (self.contentView != nil) {
        self.contentView.frame = sideBarFrame;
    }
    self.translucentView.frame = sideBarFrame;
}

- (void)move:(CGFloat)deltaFromStartX {
	CGRect sideBarFrame = self.translucentView.frame;
	CGFloat parentWidth = self.view.bounds.size.width;

	if (self.showFromRight) {
		CGFloat x = deltaFromStartX;
		if (deltaFromStartX >= self.sideBarWidth) {
			x = self.sideBarWidth;
		}
		sideBarFrame.origin.x = parentWidth - x;
	}
	else {
		CGFloat x = deltaFromStartX - _sideBarWidth;
		if (x >= 0) {
			x = 0;
		}
		sideBarFrame.origin.x = x;
	}

	if (self.contentView != nil) {
		self.contentView.frame = sideBarFrame;
	}
	self.translucentView.frame = sideBarFrame;
}

- (void)showAnimatedFrom:(BOOL)animated deltaX:(CGFloat)deltaXFromStartXToEndX {
	if ([self.delegate respondsToSelector:@selector(sideBar:willAppear:)]) {
		[self.delegate sideBar:self willAppear:animated];
	}

	CGRect sideBarFrame = self.translucentView.frame;
	CGFloat parentWidth = self.view.bounds.size.width;

	sideBarFrame.origin.x = self.showFromRight ? parentWidth - sideBarFrame.size.width : 0;

	void (^animations)() = ^{
		if (self.contentView != nil) {
			self.contentView.frame = sideBarFrame;
		}

		self.translucentView.frame = sideBarFrame;
	};
	void (^completion)(BOOL) = ^(BOOL finished)
	{
		_hasShown = YES;
		if (finished && [self.delegate respondsToSelector:@selector(sideBar:didAppear:)]) {
			[self.delegate sideBar:self didAppear:animated];
		}
	};

	if (animated) {
		[UIView animateWithDuration:self.animationDuration delay:0 options:kNilOptions animations:animations completion:completion];
	}
	else {
		animations();
		completion(YES);
	}
}

#pragma mark - Dismiss
- (void)dismiss {
	[self dismissAnimated:YES];
}

- (void)dismissAnimated:(BOOL)animated {
	if ([self.delegate respondsToSelector:@selector(sideBar:willDisappear:)]) {
		[self.delegate sideBar:self willDisappear:animated];
	}

	void (^completion)(BOOL) = ^(BOOL finished)
	{
		[self removeFromParentViewControllerCallingAppearanceMethods:YES];
		_hasShown = NO;
		self.isCurrentPanGestureTarget = NO;
		if ([self.delegate respondsToSelector:@selector(sideBar:didDisappear:)]) {
			[self.delegate sideBar:self didDisappear:animated];
		}
	};

	if (animated) {
		CGRect sideBarFrame = self.translucentView.frame;
		CGFloat parentWidth = self.view.bounds.size.width;
		sideBarFrame.origin.x = self.showFromRight ? parentWidth : -self.sideBarWidth;
		[UIView animateWithDuration:self.animationDuration
		                      delay:0
		                    options:UIViewAnimationOptionBeginFromCurrentState
		                 animations: ^{
		    if (self.contentView != nil) {
		        self.contentView.frame = sideBarFrame;
			}
		    self.translucentView.frame = sideBarFrame;
		}

		                 completion:completion];
	}
	else {
		completion(YES);
	}
}

#pragma mark - Dismiss by Pangesture
- (void)dismissAnimated:(BOOL)animated deltaX:(CGFloat)deltaXFromStartXToEndX {
	if ([self.delegate respondsToSelector:@selector(sideBar:willDisappear:)]) {
		[self.delegate sideBar:self willDisappear:animated];
	}

	void (^completion)(BOOL) = ^(BOOL finished)
	{
		[self removeFromParentViewControllerCallingAppearanceMethods:YES];
		_hasShown = NO;
		self.isCurrentPanGestureTarget = NO;
		if ([self.delegate respondsToSelector:@selector(sideBar:didDisappear:)]) {
			[self.delegate sideBar:self didDisappear:animated];
		}
	};

	if (animated) {
		CGRect sideBarFrame = self.translucentView.frame;
		CGFloat parentWidth = self.view.bounds.size.width;
		sideBarFrame.origin.x = self.showFromRight ? parentWidth : -self.sideBarWidth;

		[UIView animateWithDuration:self.animationDuration
		                      delay:0
		                    options:UIViewAnimationOptionBeginFromCurrentState
		                 animations: ^{
		    if (self.contentView != nil) {
		        self.contentView.frame = sideBarFrame;
			}
		    self.translucentView.frame = sideBarFrame;
		}

		                 completion:completion];
	}
	else {
		completion(YES);
	}
}

#pragma mark - Gesture Handler
- (void)handleTapGesture:(UITapGestureRecognizer *)recognizer {
	CGPoint location = [recognizer locationInView:self.view];
	if (!CGRectContainsPoint(self.translucentView.frame, location)) {
		[self dismissAnimated:YES];
	}
}

- (void)handlePanGesture:(UIPanGestureRecognizer *)recognizer {
	if (!self.isCurrentPanGestureTarget) {
		return;
	}

	if (recognizer.state == UIGestureRecognizerStateBegan) {
		self.panStartPoint = [recognizer locationInView:self.view];
	}

	if (recognizer.state == UIGestureRecognizerStateChanged) {
		CGPoint currentPoint = [recognizer locationInView:self.view];
		if (!self.showFromRight) {
			[self move:self.sideBarWidth + currentPoint.x - self.panStartPoint.x];
		}
		else {
			[self move:self.sideBarWidth + self.panStartPoint.x - currentPoint.x];
		}
	}

	if (recognizer.state == UIGestureRecognizerStateEnded) {
		CGPoint endPoint = [recognizer locationInView:self.view];

		if (!self.showFromRight) {
			if (self.panStartPoint.x - endPoint.x < self.sideBarWidth / 3) {
				[self showAnimatedFrom:YES deltaX:endPoint.x - self.panStartPoint.x];
			}
			else {
				[self dismissAnimated:YES deltaX:endPoint.x - self.panStartPoint.x];
			}
		}
		else {
			if (self.panStartPoint.x - endPoint.x >= self.sideBarWidth / 3) {
				[self showAnimatedFrom:YES deltaX:self.panStartPoint.x - endPoint.x];
			}
			else {
				[self dismissAnimated:YES deltaX:self.panStartPoint.x - endPoint.x];
			}
		}
	}
}

- (void)handlePanGestureToShow:(UIPanGestureRecognizer *)recognizer inView:(UIView *)view {
	if (!self.isCurrentPanGestureTarget) {
		return;
	}

	if (recognizer.state == UIGestureRecognizerStateBegan) {
		self.panStartPoint = [recognizer locationInView:view];
		[self startShow:self.panStartPoint.x];
	}

	if (recognizer.state == UIGestureRecognizerStateChanged) {
		CGPoint currentPoint = [recognizer locationInView:view];
		if (!self.showFromRight) {
			[self move:currentPoint.x - self.panStartPoint.x];
		}
		else {
			[self move:self.panStartPoint.x - currentPoint.x];
		}
	}

	if (recognizer.state == UIGestureRecognizerStateEnded) {
		CGPoint endPoint = [recognizer locationInView:view];

		if (!self.showFromRight) {
			if (endPoint.x - self.panStartPoint.x >= self.sideBarWidth / 3) {
				[self showAnimatedFrom:YES deltaX:endPoint.x - self.panStartPoint.x];
			}
			else {
				[self dismissAnimated:YES deltaX:endPoint.x - self.panStartPoint.x];
			}
		}
		else {
			if (self.panStartPoint.x - endPoint.x >= self.sideBarWidth / 3) {
				[self showAnimatedFrom:YES deltaX:self.panStartPoint.x - endPoint.x];
			}
			else {
				[self dismissAnimated:YES deltaX:self.panStartPoint.x - endPoint.x];
			}
		}
	}
}

-(void)handlePanGestureToShow:(UIPanGestureRecognizer *)recognizer inViewController:(UIViewController *)controller {
    if (!self.isCurrentPanGestureTarget) {
        return;
    }
    
    if (recognizer.state == UIGestureRecognizerStateBegan) {
        self.panStartPoint = [recognizer locationInView:controller.view];
        [self startShowInViewController:controller startX:self.panStartPoint.x];
    }
    
    if (recognizer.state == UIGestureRecognizerStateChanged) {
        CGPoint currentPoint = [recognizer locationInView:controller.view];
        if (!self.showFromRight) {
            [self move:currentPoint.x - self.panStartPoint.x];
        }
        else {
            [self move:self.panStartPoint.x - currentPoint.x];
        }
    }
    
    if (recognizer.state == UIGestureRecognizerStateEnded) {
        CGPoint endPoint = [recognizer locationInView:controller.view];
        
        if (!self.showFromRight) {
            if (endPoint.x - self.panStartPoint.x >= self.sideBarWidth / 3) {
                [self showAnimatedFrom:YES deltaX:endPoint.x - self.panStartPoint.x];
            }
            else {
                [self dismissAnimated:YES deltaX:endPoint.x - self.panStartPoint.x];
            }
        }
        else {
            if (self.panStartPoint.x - endPoint.x >= self.sideBarWidth / 3) {
                [self showAnimatedFrom:YES deltaX:self.panStartPoint.x - endPoint.x];
            }
            else {
                [self dismissAnimated:YES deltaX:self.panStartPoint.x - endPoint.x];
            }
        }
    }
}

- (BOOL)gestureRecognizer:(UIGestureRecognizer *)gestureRecognizer shouldReceiveTouch:(UITouch *)touch {
	if (touch.view != gestureRecognizer.view) {
		return NO;
	}
	return YES;
}

#pragma mark - ContentView
- (void)setContentViewInSideBar:(UIView *)contentView {
	if (self.contentView != nil) {
		[self.contentView removeFromSuperview];
	}

	self.contentView = contentView;
	self.contentView.backgroundColor = [UIColor clearColor];
	[self.view addSubview:self.contentView];
}

#pragma mark - Helper
- (void)addToParentViewController:(UIViewController *)parentViewController callingAppearanceMethods:(BOOL)callAppearanceMethods {
	if (self.parentViewController != nil) {
		[self removeFromParentViewControllerCallingAppearanceMethods:callAppearanceMethods];
	}

	if (callAppearanceMethods) [self beginAppearanceTransition:YES animated:NO];
	[parentViewController addChildViewController:self];
	[parentViewController.view addSubview:self.view];
	[self didMoveToParentViewController:self];
	if (callAppearanceMethods) [self endAppearanceTransition];
}

- (void)removeFromParentViewControllerCallingAppearanceMethods:(BOOL)callAppearanceMethods {
	if (callAppearanceMethods) [self beginAppearanceTransition:NO animated:NO];
	[self willMoveToParentViewController:nil];
	[self.view removeFromSuperview];
	[self removeFromParentViewController];
	if (callAppearanceMethods) [self endAppearanceTransition];
}

@end

File: CDRTranslucentSideBar/CDRTranslucentSideBar/CDRTranslucentSideBar.swift
//
//  CDRTranslucentSideBar.swift
//  CDRTranslucentSideBar
//
//  Created by UetaMasamichi on 2014/12/28.
//  Copyright (c) 2014年 nscallop. All rights reserved.
//

import UIKit

@objc public protocol CDRTranslucentSideBarDelegate {
    optional func sideBar(sideBar: CDRTranslucentSideBar, didAppear animated: Bool)
    optional func sideBar(sideBar: CDRTranslucentSideBar, willAppear animated: Bool)
    optional func sideBar(sideBar: CDRTranslucentSideBar, didDisappear animated: Bool)
    optional func sideBar(sideBar: CDRTranslucentSideBar, willDisappear animated: Bool)
}

public enum CDRTranslucentSideBarStyle: Int {
    case ExtraLight
    case Light
    case Dark
}

private let defaultSideBarWidth: CGFloat = 200
private let defaultAnimationDuration: Double = 0.25

public class CDRTranslucentSideBar: UIViewController {
    
    //MARK: - Properties
    public weak var delegate: CDRTranslucentSideBarDelegate?
    public let sideBarWidth: CGFloat
    public let animationDuration: Double
    public var isCurrentPanGestureTarget: Bool
    public var tag: UInt
    public var style: CDRTranslucentSideBarStyle
    
    private var _contentView: UIView?
    public var contentView: UIView? {
        get {
            return self._contentView
        }
        set {
            self._contentView?.removeFromSuperview()
            self._contentView = newValue
            self._contentView?.backgroundColor = UIColor.clearColor()
            
            if let contentView = self._contentView {
                self.view.addSubview(contentView)
            }
        }
    }
    
    private(set) var hasShown: Bool
    private(set) var showFromRight: Bool

    //MARK: - Private Members
    private var translucentView: UIView?
    private var tapGestureRecognizer: UITapGestureRecognizer?
    private var panGestureRecognizer: UIPanGestureRecognizer?
    private var panStartPoint: CGPoint
    
    //MARK: - Initializer
    public override convenience init () {
        self.init(style: .Dark, showFromRight: false, width: defaultSideBarWidth, animationDuration: defaultAnimationDuration)
    }
    
    public convenience init(style: CDRTranslucentSideBarStyle) {
        self.init(style: style, showFromRight: false, width: defaultSideBarWidth, animationDuration: defaultAnimationDuration)
    }
    
    public convenience init(style: CDRTranslucentSideBarStyle, showFromRight: Bool){
        self.init(style: style, showFromRight: showFromRight, width: defaultSideBarWidth, animationDuration: defaultAnimationDuration)
    }
    
    //Custom Initializer
    public convenience init(style: CDRTranslucentSideBarStyle, showFromRight: Bool, width: CGFloat) {
        self.init(style: style, showFromRight: showFromRight, width: defaultSideBarWidth, animationDuration: defaultAnimationDuration)
    }
    
    public init(style: CDRTranslucentSideBarStyle, showFromRight: Bool, width: CGFloat, animationDuration: Double) {
        self.style = style
        self.showFromRight = showFromRight
        self.sideBarWidth = width
        self.animationDuration = animationDuration
        self.isCurrentPanGestureTarget = false
        self.tag = 0
        self.hasShown = false
        self.panStartPoint = CGPointMake(0, 0)
        super.init()
    }

    public required init(coder aDecoder: NSCoder) {
        self.style = .Dark
        self.showFromRight = false
        self.sideBarWidth = defaultSideBarWidth
        self.animationDuration = defaultAnimationDuration
        self.isCurrentPanGestureTarget = false
        self.tag = 0
        self.hasShown = false
        self.panStartPoint = CGPointMake(0, 0)
        super.init(coder: aDecoder)
        
    }
    
    public override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: NSBundle?) {
        self.style = .Dark
        self.showFromRight = false
        self.sideBarWidth = defaultSideBarWidth
        self.animationDuration = defaultAnimationDuration
        self.isCurrentPanGestureTarget = false
        self.tag = 0
        self.hasShown = false
        self.panStartPoint = CGPointMake(0, 0)
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }
    
    //MARK: - View Life Cycle
    public override func viewDidLoad() {
        
        //Create TranslucentView
        if UIDevice.currentDevice().systemVersion.compare("8.0", options: NSStringCompareOptions.NumericSearch) != NSComparisonResult.OrderedAscending {
            //for iOS 8 or later
            let blurStyle: UIBlurEffectStyle = UIBlurEffectStyle(rawValue: self.style.rawValue)!
            let blurEffect = UIBlurEffect(style: blurStyle)
            let visualEffectView = UIVisualEffectView(effect: blurEffect)
            let frame = CGRectMake(self.showFromRight ? self.view.bounds.size.width : -self.sideBarWidth, 0, self.sideBarWidth, self.view.bounds.size.height)
            visualEffectView.frame = frame
            visualEffectView.contentMode = self.showFromRight ? .TopRight : .TopLeft
            visualEffectView.clipsToBounds = true
            self.translucentView = visualEffectView
        } else {
            //for iOS 7
            let frame = CGRectMake(self.showFromRight ? self.view.bounds.size.width : -self.sideBarWidth, 0, self.sideBarWidth, self.view.bounds.size.height)
            let toolbar = UIToolbar(frame: frame)
            let barStyle: UIBarStyle = self.style == .Dark ? .Black : .Default
            toolbar.barStyle = barStyle
            toolbar.contentMode = self.showFromRight ? .TopRight : .TopLeft
            toolbar.clipsToBounds = true
            self.translucentView = toolbar
        }
        
        self.view.layer.insertSublayer(self.translucentView!.layer, atIndex: 0)
        
        self.view.backgroundColor = UIColor.clearColor()
        self.tapGestureRecognizer = UITapGestureRecognizer(target: self, action: "handleTapGesture")
        self.tapGestureRecognizer!.delegate = self
        self.view.addGestureRecognizer(self.tapGestureRecognizer!)
        self.panGestureRecognizer = UIPanGestureRecognizer(target: self, action: "handlePanGesture")
        self.panGestureRecognizer!.minimumNumberOfTouches = 1
        self.panGestureRecognizer!.maximumNumberOfTouches = 1
        self.view.addGestureRecognizer(self.panGestureRecognizer!)
    }
    
    
    //MARK: - Layout
    public override func shouldAutorotate() -> Bool {
        return true
    }
    
    public override func willAnimateRotationToInterfaceOrientation(toInterfaceOrientation: UIInterfaceOrientation, duration: NSTimeInterval) {
        super.willAnimateRotationToInterfaceOrientation(toInterfaceOrientation, duration: duration)
        
        if self.isViewLoaded() && self.view.window != nil {
            self.layoutSubViews()
        }
    }
    
    private func layoutSubViews() {
        if let parent = self.parentViewController {
            let x = self.showFromRight ? parent.view.bounds.width - self.sideBarWidth : 0
            
            self.contentView?.frame = CGRectMake(x, 0, self.sideBarWidth, parent.view.bounds.size.height)
        }
    }
    
    //MARK: - Helper
    private func addToParentController(parentViewController:UIViewController?, callAppearanceMethods: Bool) {
        if self.parentViewController != nil {
            self.removeFromParentviewController(callAppearanceMethods)
        }
        
        if callAppearanceMethods {
            self.beginAppearanceTransition(true, animated: false)
        }
        
        parentViewController?.addChildViewController(self)
        parentViewController?.view.addSubview(self.view)
        self.didMoveToParentViewController(self)
        
        if callAppearanceMethods {
            self.endAppearanceTransition()
        }
    }
    
    private func removeFromParentviewController(callAppearanceMethods: Bool) {
        if callAppearanceMethods {
            self.beginAppearanceTransition(false, animated: false)
        }
        
        self.willMoveToParentViewController(nil)
        self.view.removeFromSuperview()
        self.removeFromParentViewController()
        
        if callAppearanceMethods {
            self.endAppearanceTransition()
        }
        
    }
    
    
    //MARK: - Show
    public func show() {
        self.show(true)
    }
    
    public func show(animated: Bool) {
        let rootViewController = UIApplication.sharedApplication().keyWindow?.rootViewController
        var controller: UIViewController?
        while rootViewController?.presentedViewController != nil {
            controller = rootViewController?.presentedViewController
        }
        
        if let parentController = controller {
            self.show(inViewController: parentController, animated: animated)
        }
    }
    
    public func show(inViewController controller: UIViewController) {
        self.show(inViewController: controller, animated: true)
    }
    
    public func show(inViewController controller: UIViewController, animated: Bool) {
        
        self.delegate?.sideBar?(self, willAppear: true)
        
        self.addToParentController(controller, callAppearanceMethods: true)
        self.view.frame = controller.view.bounds
        
        let parentWidth = self.view.bounds.size.width
        let startSideBarFrame = CGRectMake(self.showFromRight ? parentWidth : -self.sideBarWidth, self.view.bounds.origin.y, self.sideBarWidth, self.view.bounds.size.height)
        
        self.contentView?.frame = startSideBarFrame
        
        let stopSideBarFrame = CGRectMake(self.showFromRight ? parentWidth - self.sideBarWidth : 0, startSideBarFrame.origin.y, startSideBarFrame.size.width, startSideBarFrame.size.height)
        
        let animations: () -> () = {
            self.contentView?.frame = stopSideBarFrame
            self.translucentView?.frame = stopSideBarFrame
        }
        
        let completion: (Bool) -> () = {(finished: Bool) in
            self.hasShown = true
            self.isCurrentPanGestureTarget = true
            
            if finished {
                self.delegate?.sideBar?(self, didAppear: true)
            }
        }
        
        if animated {
            UIView.animateWithDuration(NSTimeInterval(self.animationDuration),
                animations: animations,
                completion: completion)
        } else {
            animations()
            completion(true)
        }
    }
    
    //MARK: - Dismiss
    public func dismiss() {
        self.dismiss(true)
    }
    
    public func dismiss(animated: Bool) {
        self.delegate?.sideBar?(self, willDisappear: true)
        
        let parentWidth = self.view.bounds.size.width
        let stopSideBarFrame = CGRectMake(self.showFromRight ? parentWidth  : -self.sideBarWidth, self.view.bounds.origin.y, self.view.bounds.size.width, self.view.bounds.size.height)
        
        let animations: () -> () = {
            self.contentView?.frame = stopSideBarFrame
            self.translucentView?.frame = stopSideBarFrame
        }
        
        let completion: (Bool) -> () = { (finished: Bool) in
            self.removeFromParentviewController(true)
            self.hasShown = false
            self.isCurrentPanGestureTarget = false
            self.delegate?.sideBar?(self, didDisappear: true)
        }
        
        if animated {
            UIView.animateWithDuration(NSTimeInterval(self.animationDuration),
                animations: animations,
                completion: completion)
        } else {
            animations()
            completion(true)
        }
    }
    
    //MARK: - Gesture Handler
    private func handleTapGesture(recognizer: UITapGestureRecognizer) {
        let location = recognizer.locationInView(self.view)
        
        if let frame = self.translucentView?.frame {
            if !CGRectContainsPoint(frame, location) {
                            self.dismiss(true)
            }
        }
    }
    
    private func handlePanGesture(recognizer: UIPanGestureRecognizer) {
        if !self.isCurrentPanGestureTarget {
            return
        }
        
        if recognizer.state == .Began {
            self.panStartPoint = recognizer.locationInView(self.view)
        }
        
        if recognizer.state == .Changed {
            let currentPoint = recognizer.locationInView(self.view)
            
            if !self.showFromRight {
                
            } else {
                
            }
        }
        
        //if recognizer.state = .Ended {
            
        //}
    }
    

    //MARK: - Show by PanGesture
    private func startToShow(startX: CGFloat) {
        
    }
   
}

extension CDRTranslucentSideBar: UIGestureRecognizerDelegate {
    
}
